### PR TITLE
Fix #1064 (let* tail if/else) and #1065 (Array::find) codegen bugs; Perceus elision + eval harnesses

### LINKED ISSUE #1064: `let* <bind> = <Option expr>` followed by if/else tail traps with "unreachable"
## 修正 (2026-07-22)

原因は `let*`/`?` railway-bind の lowering (`lib/@vibe/compiler/codegen/common_base/desugar_trait_dict.vibe` の `railway_rw`/`rewrite_expr` 内、`__try_bind` sentinel の処理) にあった。

`let* rv = half(x)` は `EMatch(half(x), [(Some(rv) => rest), (None => None)])` に desugar されるが、この際 `extend_var_types(var_types, name, inferred)` という呼び出しで `rv`(bind された名前) の型として `inferred`(= `half(x)` という**式全体**の型、つまり `"Option"`) を記録していた。しかし `rv` が実際に束縛されるのは Option の**中身**（unwrap された payload）であって、Option 値そのものではない — wrapper 型とpayload型を混同するバグだった。

この結果、`rest` の中で `rv == 0` のような比較を行うと、`infer_arg_type_name` が `var_types` から `rv` の型を `"Option"` と(誤って)引いてしまい、`EBinOp` の `==` 処理 (`compile_call.vibe`/`desugar_trait_dict.vibe`) が「Option 同士の構造比較」と誤判定して `opt_eq_match` 経路に迂回させていた。実際には `rv` はただの `Int` なので、この構造比較は生スカラー値をヒープオブジェクトであるかのように読み込み、意味のない結果を返していた（`rv == <任意の値>` が常に `true` になる、など）。最終的に `assert(f(4) == Some(2))` が(誤った計算結果により)失敗し、`assert` 失敗は `unreachable` trap にコンパイルされる — これが報告された trap の正体。

同じ挙動を素の `match` で手書きした場合は再現しないことを確認済み — `let*` sugar 特有の `var_types` 汚染が原因と確定。

### 修正内容

`extend_var_types(var_types, name, inferred)` の呼び出し（`railway_rw`/`rewrite_expr` の両方）を削除し、bind された名前を `var_types` に登録しないようにした。`fn_returns` テーブルは呼び出し先の戻り値型の「先頭型名」しか持たない (`"Option"` であって `"Option[Int]"` ではない) ため、payload の正確な型はこの時点では復元できない — 追跡しないことで `infer_arg_type_name` が `None`(型不明)を返し、素の scalar `==` 比較にフォールバックする、という安全側に倒す修正。

### 検証

- issue 本文の最小再現 — pass (`f(4) == Some(2)`)
- 追加確認: `rv == 0`/`rv == 2`/`rv == 999999` それぞれの正しい真偽値、chained `let*` (`checker_railway_mixing_test.vibe` の同モナドチェーン型検査テスト)
- `scripts/compiler_gate.sh` (bundle sync / module-source sync / seed→stage1→stage2→stage3 self-host fixpoint / RC bootstrap / 45件の回帰チェック) — 全て pass、fixpoint (`stage2==stage3`) 確認済み
- `lib/@vibe/compiler/tests/checker_railway_mixing_test.vibe` を含むフルスイート — pass 確認

コミット: `1578147` (branch `claude/almide-crates-benchmark-sqmcfq`)

### LINKED ISSUE #1065: Array::find traps (unreachable); related closure-capture pattern corrupts RC heap (memory access out of bounds)
## 修正 (2026-07-22)

原因は当初疑っていた「クロスファイル export-rename 後の qualified-sugar 呼び出しが未解決」ではなく、`Array::find` 専用のハンドライト codegen (`lib/@vibe/compiler/codegen/expr/compile_call.vibe` の `fname == "Array::find"` 分岐) 自体のバグだった。

`Array::any`/`Array::all` は `Bool` という生スカラーを返すので `res_l` に raw な 0/1 を積んでも正しいが、`Array::find` は `Option[T]` を返す — Option の値は tag + fields を持つ実ヒープオブジェクトで、`None`/`Some(x)` は必ずヒープ確保を経由する (`compile_expr.vibe` のベア `EIdent("None", ...)` ctor 経路や `compile_call.vibe` の一般 ctor-call 経路を参照)。ところが `Array::find` の実装は「未発見」を `i64.const 0` に、「発見」を要素の生値そのものに `res_l` へ直接書き込んでいた — 返る「Option」は tag もヒープオブジェクトも持たない、ただの生スカラーだった。

`assert(Array::find(..) == Some(x))` はこの構造的に壊れた値を比較して失敗し、`assert` の失敗は wasm `unreachable` にコンパイルされる — これが報告された trap の正体（名前解決の失敗ではなく、値の構築ミス）。

### 修正内容

`res_l` への書き込みを、`None`/`Some(elem)` を合成した ctor-call AST に対して `ce` (compile_expr の再帰呼び出し) を通す方式に変更 — リテラルの `Some(x)`/`None` 式が通るのと同じ経路を使うことで、RC の有無に関わらず正しい ctor 構築コードが生成される。

### 検証

- issue 本文の最小再現 (import あり/なし、bare 呼び出し/qualified sugar 呼び出し) — 全て pass
- 追加エッジケース: not-found (`None` 分岐)、ヒープ確保が必要な `String` 要素、先頭一致優先セマンティクス — 全て pass
- `scripts/compiler_gate.sh` (bundle sync / module-source sync / seed→stage1→stage2→stage3 self-host fixpoint / RC bootstrap / 45件の回帰チェック) — 全て pass、fixpoint (`stage2==stage3`) 確認済み
- `lib/@vibe/compiler/tests/` フルスイート — 実行中に主要ファイル群 (checker/codegen/compiler 系) は全て pass 確認

コミット: `1578147` (branch `claude/almide-crates-benchmark-sqmcfq`)

### PR DESCRIPTION
**Pull request by mizchi**
## Summary

- **Fixes #1064**: `let*`/`?` railway-bind desugar recorded the bound name's type as the *wrapper* type (`"Option"`/`"Result"`) instead of leaving it untracked. A later `==` comparison against the bound (unwrapped scalar) name was then misclassified as a structural Option/Result comparison and computed garbage — `let* rv = half(x)` followed by `if rv == 0 {...}` silently corrupted the branch taken. Root-caused down to a single `extend_var_types` call in `desugar_trait_dict.vibe`'s `railway_rw`/`rewrite_expr`.
- **Fixes #1065**: the hand-written inline codegen for `Array::find` in `compile_call.vibe` stored a raw scalar (`i64.const 0` for "not found", the bare element for "found") into the result slot instead of a properly constructed `Option` heap object (`None`/`Some(elem)`). Any `== Some(x)` check against the result was then comparing structurally bogus values — not a name-resolution issue as originally suspected, a value-construction bug in the `Array::find` fast path specifically (`Array::any`/`Array::all` are unaffected since they return a plain scalar `Bool`).
- Perceus RC: elide dead-alias dup/drop pairs (occurrence-local subset of almide's `alias_safety.rs` fixpoint analysis) — carried over from #1056 Phase B.
- `bench/binary_size/`: 5-program wasm binary-size regression bench (RC on/off) — Phase C from #1056.
- `eval/msr/` + `eval/lang-bench/`: Modification Survival Rate and cross-language same-model-snapshot evaluation harnesses (almide-inspired), plus one completed round of results (`2026-07-22-r1`) for both.
- `docs/pl-survey-2026-07.md` / `docs/BENCHMARKS.md` / `docs/spec/rc-port.md` updates recording the above.

Note: this branch's first commit (almide crates survey doc) was already squash-merged into `main` as #1055/8532ccd; this PR is the rebased continuation carrying everything after that.

## Verification

- `scripts/compiler_gate.sh` (bundle sync, module-source sync, seed→stage1→stage2→stage3 self-hosting fixpoint, RC self-compile bootstrap, 45 regression checks) — full pass on the rebased tree.
- `lib/@vibe/compiler/tests/` full suite — pass.
- #1064/#1065 minimal repros from both issues, plus additional edge cases (None branch, heap-allocated `String` elements, chained `let*`, first-match semantics) — all pass.
- Perceus: `perceus_rc_test.vibe` (incl. 2 new regression tests), `codegen_heap_e2e_test.vibe`, `scripts/verify_rc.sh` (RC self-compile byte-identical), `fixtures/rc_*_test.vibe`.

## Test plan
- [x] `scripts/compiler_gate.sh`
- [x] `lib/@vibe/compiler/tests/` full suite
- [x] #1064/#1065 repros + edge cases
- [x] `perceus_rc_test.vibe` / `codegen_heap_e2e_test.vibe` / `scripts/verify_rc.sh`
- [x] `bash scripts/bench_binary_size.sh`
- [x] `eval/msr/run_msr.sh` and `eval/lang-bench/acceptance_test.sh` (real rounds, not mocked)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019BXLigdgi5CSrdUPqdaa39

---
_Generated by [Claude Code](https://claude.ai/code/session_019BXLigdgi5CSrdUPqdaa39)_

### CODE REVIEW

File: lib/@vibe/compiler/codegen/common_base/desugar_trait_dict.vibe [hunk 1]
```diff
@@ -2339,7 +2358,24 @@ fn railway_rw(expr: Expr, struct_sets: Array[(String, Array[String])], var_types
...
+            // silently computing nonsense (any comparison against a plain
+            // scalar payload behaved like it were reading through an actual
+            // Option/Result value's heap layout). No replacement type is
+            // recorded: fn_returns only carries the head name of a call's
+            // return type ("Option", not "Option[Int]"), so the payload's
+            // real type isn't recoverable here -- leaving `name` untracked
+            // falls through infer_arg_type_name to `None` (unknown type),
+            // which keeps the plain scalar `==`/`!=` codegen path, correct
+            // for the common case this railway-bind sugar exists for.
+            let body_vars = var_types
```
[the remark this answers is not in the document]
Comment (mizchi) [reply lib/@vibe/compiler/codegen/common_base/desugar_trait_dict.vibe]: Good catch, thanks — verified this was a real regression (`let* outer = Some(Some(1)); let* inner = outer` failed to compile with "unknown name: Err"). Fixed in 86f46b4: added `railway_bind_payload_type` to recover the payload's head type specifically for the literal `Some(x)`/`Ok(x)` nesting shape, while leaving every other `val` shape (plain function call, bare identifier) untracked so the original #1064 scalar-payload fix still holds. Verified both the nested-Option and nested-Result cases now pass, alongside the original #1064 repro, `checker_railway_mixing_test.vibe`, and the full `compiler_gate.sh`.

---
_Generated by [Claude Code](https://claude.ai/code)_
